partition_snapshot_row_cursor.hh

```diff
@@ -111,52 +111,94 @@ class partition_snapshot_row_cursor final {
-        };
+        bool unique_owner = false;
+        is_continuous continuous = is_continuous::no; // Range continuity in the direction of lower keys (in cursor schema domain).
     };
 
-    const schema& _schema;
+    const schema& _schema; // query domain
     partition_snapshot& _snp;
-    utils::small_vector<position_in_version, 2> _heap;
+
+    // _heap contains iterators which are ahead of the cursor.
+    // _current_row contains iterators which are directly below the cursor.
+    utils::small_vector<position_in_version, 2> _heap; // query domain order
     utils::small_vector<position_in_version, 2> _current_row;
+
+    // For !_reversed cursors points to the entry which
+    // is the lower_bound() of the current position in table schema order.
+    // For _reversed cursors it can be either lower_bound() in table order
+    // or lower_bound() in cursor's order, so should not be relied upon.
+    // if current entry is in the latest version then _latest_it points to it,
+    // also in _reversed mode.
     std::optional<mutation_partition::rows_type::iterator> _latest_it;
+
+    // Continuity corresponding to ranges which are not represented in _heap because the cursor
+    // went pass all the entries in those versions.
+    bool _background_continuity = false;
+
     bool _continuous{};
     bool _dummy{};
     const bool _unique_owner;
-    position_in_partition _position;
+    const bool _reversed;
+    position_in_partition _position; // table domain
     partition_snapshot::change_mark _change_mark;
 
+    position_in_partition_view to_table_domain(position_in_partition_view pos) const {
+        if (_reversed) [[unlikely]] {
+            return pos.reversed();
+        }
+        return pos;
+    }
+
+    position_in_partition_view to_query_domain(position_in_partition_view pos) const {
+        if (_reversed) [[unlikely]] {
+            return pos.reversed();
+        }
+        return pos;
+    }
+
+    struct version_heap_less_compare {
+        rows_entry::tri_compare _cmp;
+        partition_snapshot_row_cursor& _cur;
+    public:
+        explicit version_heap_less_compare(partition_snapshot_row_cursor& cur)
+            : _cmp(cur._schema)
+            , _cur(cur)
+        { }
+
+        bool operator()(const position_in_version& a, const position_in_version& b) {
+            auto res = _cmp(_cur.to_query_domain(a.it->position()), _cur.to_query_domain(b.it->position()));
+            return res > 0 || (res == 0 && a.version_no > b.version_no);
+        }
+    };
+
     // Removes the next row from _heap and puts it into _current_row
     bool recreate_current_row() {
         _current_row.clear();
+        _continuous = _background_continuity;
+        _dummy = true;
         if (_heap.empty()) {
+            if (_reversed) {
+                _position = position_in_partition::before_all_clustered_rows();
+            } else {
+                _position = position_in_partition::after_all_clustered_rows();
```

Comment:
This line affects only non-reversed reads. Was it a bug that it was missing previously?

Reply:
There was no bug. This change is in order to support the new, more relaxed, semantics for maybe_advance_to():

>  // If no such entry exists, the cursor is positioned at an extreme position in the direction of
>     // the cursor (min for reversed cursor, max for forward cursor) and not pointing at a row
>     // but still valid.
>    

It now allows advancing even if there is no dummy entry. Forward reads don't rely on it yet, but it's good to be symmetric.